relayer/strategies.go

```diff
@@ -1,58 +1,241 @@
 	"time"
+
+	"github.com/avast/retry-go"
+	"github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
 )
 
-// StartRelayer starts the main relaying loop
+// ActiveChannel represents an IBC channel and whether there is an active goroutine relaying packets against it.
+type ActiveChannel struct {
+	channel *types.IdentifiedChannel
+	active  bool
+}
+
+// StartRelayer starts the main relaying loop.
 func StartRelayer(src, dst *Chain, maxTxSize, maxMsgLength uint64) (func(), error) {
 	doneChan := make(chan struct{})
+	channels := make(chan *ActiveChannel, 10)
+	var srcOpenChannels []*ActiveChannel
+
 	go func() {
 		for {
 			select {
 			case <-doneChan:
 				return
 			default:
-				// Fetch any unrelayed sequences depending on the channel order
-				sp, err := UnrelayedSequences(src, dst)
+				// Query the list of channels on the src connection
+				srcChannels, err := QueryChannelsOnConnection(src)
 				if err != nil {
-					src.Log(fmt.Sprintf("unrelayed sequences error: %s", err))
-				} else {
-					if len(sp.Src) > 0 && src.debug {
-						src.Log(fmt.Sprintf("[%s] unrelayed-packets-> %v", src.ChainID(), sp.Src))
-					}
-					if len(sp.Dst) > 0 && dst.debug {
-						dst.Log(fmt.Sprintf("[%s] unrelayed-packets-> %v", dst.ChainID(), sp.Dst))
-					}
-					if !sp.Empty() {
-						if err = RelayPackets(src, dst, sp, maxTxSize, maxMsgLength); err != nil {
-							src.Log(fmt.Sprintf("relay packets error: %s", err))
-						}
-					}
+					fmt.Printf("error querying all channels on chain{%s}@connection{%s}: %v \n",
+						src.ChainID(), src.ConnectionID(), err)
+
+					os.Exit(1)
 				}
 
-				// Fetch any unrelayed acks depending on the channel order
-				ap, err := UnrelayedAcknowledgements(src, dst)
-				if err != nil {
-					src.Log(fmt.Sprintf("unrelayed acks error: %s", err))
-				} else {
-					if len(ap.Src) > 0 && src.debug {
-						src.Log(fmt.Sprintf("[%s] unrelayed-acks-> %v", src.ChainID(), ap.Src))
-					}
-					if len(ap.Dst) > 0 && dst.debug {
-						dst.Log(fmt.Sprintf("[%s] unrelayed-acks-> %v", dst.ChainID(), ap.Dst))
-					}
-					if !ap.Empty() {
-						if err = RelayAcknowledgements(src, dst, ap, maxTxSize, maxMsgLength); err != nil && src.debug {
-							src.Log(fmt.Sprintf("relay acks error: %s", err))
-						}
+				// Filter for open channels that are not already in our slice of open channels
+				// TODO implement a filter list of channels we want to relay against or a list of channels to ignore
+				srcOpenChannels = FilterOpenChannels(srcChannels, srcOpenChannels)
+
+				// Spin up a goroutine to relay packets & acks for each channel that isn't already being relayed against
+				for _, channel := range srcOpenChannels {
+					if !channel.active {
+						channel.active = true
+						go RelayUnrelayedPacketsAndAcks(src, dst, maxTxSize, maxMsgLength, channel, channels, doneChan)
 					}
 				}
 
-				time.Sleep(100 * time.Millisecond)
+				for channel := range channels {
+					channel.active = false
+					break
```

Comment:
This break looks out of place. Is this intended to only set the first channel inactive and then stop?

Reply:
We discussed this but leaving a comment in case anyone else is following along here.

At L53 we are waiting for one of the goroutines that are running `RelayUnrelayedPacketsAndAcks()` to write to the channel which indicates that the goroutine finished running (either encountered an error or returned). Each goroutine is associated with an IBC channel that we are relaying on so when one finishes running we want to break from this block, update the slice of open channels and re-run the entire outer for loop.